modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscriptionManager.java

```diff
@@ -274,6 +274,242 @@ private void removeLocalSubscriptionAndNotify(AndesSubscription subscription) th
+            ProtocolType protocolType, DestinationType destinationType, String bindingKeyPattern, String
+            connectedNode) throws AndesException {
+
+        Set<AndesSubscription> filteredSubscriptions = new HashSet<>();
+        String messageRouter = destinationType.getAndesMessageRouter();
+
+        Query<AndesSubscription> subscriptionQuery = and
+                (equal(AndesSubscription.DURABILITY, isDurable), equal(AndesSubscription.NODE_ID, connectedNode),
+                 equal(AndesSubscription.PROTOCOL, protocolType), equal(AndesSubscription.ROUTER_NAME, messageRouter),
+                 contains(AndesSubscription.ROUTING_KEY, bindingKeyPattern.toLowerCase()));
+
+        Iterable<AndesSubscription> subscriptions = subscriptionRegistry.exucuteQuery(subscriptionQuery);
+
+        for (AndesSubscription subscription : subscriptions) {
+            filteredSubscriptions.add(subscription);
+        }
+        return filteredSubscriptions;
+    }
+
+
+
+    /**
+     * Get inactive subscriptions filtered by identifier pattern and tokenized binding key
+     *
+     * @param bindingKeyPattern regex to match with binding key of subscriber
+     * @return Set of subscriptions filtered according to search criteria
+     * @throws AndesException
+     */
+    private Set<AndesSubscription> getInactiveSubscriptionsByTokenizedBindingKeyMatch(String bindingKeyPattern)
+            throws AndesException {
+
+        Set<AndesSubscription> filteredSubscriptions = new HashSet<>();
+
+        List<AndesSubscription> allInactiveSubscriptions = getInactiveSubscriberRepresentations();
+
+        for (AndesSubscription inactiveSubscription : allInactiveSubscriptions) {
+            if(inactiveSubscription.getStorageQueue().getMessageRouterBindingKey().toLowerCase().
+                    contains(bindingKeyPattern.toLowerCase())){
+                filteredSubscriptions.add(inactiveSubscription);
+            }
+        }
+        return filteredSubscriptions;
+    }
+
+    /**
+     * Filter inactive subscriptions
+     *
+     * @param subscriptions subscription list for further filtering
+     * @param protocolType protocol of the subscription
+     * @param destinationType type of subscription (QUEUE/TOPIC/DURABLE_TOPIC)
+     * @param subscriptionIdPattern regex to match with ID of the subscriber
+     * @param isExactMatchSubscriptionId exact match of subscription id or not
+     * @return Set of subscriptions filtered according to search criteria
+     */
+    private Set<AndesSubscription> filterInactiveSubscriptionsBySubscriptionId(Set<AndesSubscription> subscriptions,
+            ProtocolType protocolType, DestinationType destinationType,String subscriptionIdPattern,
+            boolean isExactMatchSubscriptionId){
+
+        Set<AndesSubscription> filteredSubscriptions = new HashSet<>();
+        String messageRouter = destinationType.getAndesMessageRouter();
+
```

Comment:
In this case, isExactMatchSubscriptionId condition is checked for each subscription. But that is common to the collection of subscriptions. So checking  isExactMatchSubscriptionId for each subscription is waste & it affects the performance. That's why isExactMatchSubscriptionId extracted out in an outside if condition.